File: hingeprot/extract.py
#!/usr/bin/env python3
"""
extract.py  (converted from extract.f / program hingebend)

Goal: match the provided Fortran behavior as closely as possible.

Expects files (same as Fortran):
  rescale
  coordinates
  alpha.cor
  1coor .. 10coor
  newcoordinat.mds
  slowmodes
  crosscorrslow1
  crosscorrslow2

Produces:
  anm_length
  newcoordinat2.mds
  mapping.out
  coor1.mds12 coor2.mds12 coor3.mds12 coor4.mds12
  gnm1anmvector gnm2anmvector
  hinges
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from pathlib import Path
from typing import List, Tuple


# ---------------------------- helpers ----------------------------

def _as_float(s: str, default: float = 0.0) -> float:
    try:
        return float(s)
    except Exception:
        return default


def _as_int(s: str, default: int = 0) -> int:
    try:
        return int(s)
    except Exception:
        return default


def _read_first_float(path: str | Path) -> float:
    txt = Path(path).read_text(encoding="utf-8", errors="replace").strip().split()
    if not txt:
        raise RuntimeError(f"Empty file: {path}")
    return float(txt[0])


def _read_first_int(path: str | Path) -> int:
    txt = Path(path).read_text(encoding="utf-8", errors="replace").strip().split()
    if not txt:
        raise RuntimeError(f"Empty file: {path}")
    return int(float(txt[0]))


# ---------------------------- format parsers --------------------------------
# Fortran formats:
# 97  format (17x,A3,2x,I4)     -> alpha.cor: residue type + cano
# 98  format (4x,4(3x,F8.5))    -> coor files: (skip 4 chars = index), then x y z boy
# 99  format (I4,2x,I4,10(3x,F8.5))
# 100 format (a4,8x,a4,1x,a3,1x,a1,i4,a1,f11.3,f8.3,f8.3,1x,f5.2)  (Fortran may ignore occ on read/write)
# 101 format (I4,4(3x,F8.5))
# 16  format(4x,10(4x,F8.5))    -> slowmodes rows


@dataclass
class AtomRow:
    label: str      # a4
    attyp: str      # a4
    restyp: str     # a3
    chtyp: str      # a1 (Fortran variable was char*2 but format reads 1 char)
    ind: int        # i4
    resex: str      # a1
    x: float        # f11.3
    y: float        # f8.3
    z: float        # f8.3
    occ: float = 0.0  # optional f5.2


def parse_alpha_line(line: str) -> Tuple[str, int]:
    # 97: (17x,A3,2x,I4)
    restyp = line[17:20].strip()
    cano = _as_int(line[22:26].strip(), 0)
    return restyp, cano


def parse_coor_line(line: str) -> Tuple[float, float, float, float]:
    """
    Fortran 98 reads: 4x then 4 floats.
    The file line is typically written by rbp/anm stage as:
      I4 + 4 floats (x y z boy)  -> split gives 5 tokens
    Fortran skips the first 4 chars (the index field). So we must ignore the first token if present.
    """
    parts = line.strip().split()
    if len(parts) >= 5:
        # common case: [index, x, y, z, boy]
        # verify first token is integer-ish; if not, still treat as index to mimic 4x skip
        return (float(parts[1]), float(parts[2]), float(parts[3]), float(parts[4]))
    if len(parts) == 4:
        # rare case: no index token
        return (float(parts[0]), float(parts[1]), float(parts[2]), float(parts[3]))

    # fixed-slice fallback (best effort)
    # If line has index in cols 1-4, floats likely start after that; attempt from known widths
    # This is only used if split fails.
    x = _as_float(line[7:15].strip(), 0.0)
    y = _as_float(line[18:26].strip(), 0.0)
    z = _as_float(line[29:37].strip(), 0.0)
    b = _as_float(line[40:48].strip(), 0.0)
    return (x, y, z, b)


def parse_mds100(line: str) -> AtomRow:
    """
    Parse using fixed columns for format 100.
    Note: Fortran sometimes ignores occ (f5.2) if not in I/O list,
    but the record may still contain it. We parse it if present; otherwise 0.0.
    """
    # make line long enough to slice safely
    s = line.rstrip("\n")
    s_pad = s + (" " * max(0, 80 - len(s)))

    label = s_pad[0:4].strip()
    attyp = s_pad[12:16].strip()
    restyp = s_pad[17:20].strip()
    chtyp = s_pad[21:22].strip()
    ind = _as_int(s_pad[22:26].strip(), 0)
    resex = s_pad[26:27].strip()
    x = _as_float(s_pad[27:38].strip(), 0.0)
    y = _as_float(s_pad[38:46].strip(), 0.0)
    z = _as_float(s_pad[46:54].strip(), 0.0)
    occ = _as_float(s_pad[55:60].strip(), 0.0)  # may be absent/blank
    return AtomRow(label, attyp, restyp, chtyp, ind, resex, x, y, z, occ)


def format_mds100(row: AtomRow, include_occ: bool) -> str:
    """
    Mimic Fortran format 100:
      a4,8x,a4,1x,a3,1x,a1,i4,a1,f11.3,f8.3,f8.3,(optional) 1x,f5.2
    In Fortran, if occ is not in the I/O list, the record ends after z.
    So we support include_occ=False to match write(35,100) behavior.
    """
    base = (
        f"{row.label:<4}"
        + " " * 8
        + f"{row.attyp:<4} "
        + f"{row.restyp:<3} "
        + f"{row.chtyp:<1}"
        + f"{row.ind:4d}"
        + f"{row.resex:<1}"
        + f"{row.x:11.3f}{row.y:8.3f}{row.z:8.3f}"
    )
    if include_occ:
        base += f" {row.occ:5.2f}"
    return base + "\n"


def format_101(i: int, x: float, y: float, z: float, boy: float) -> str:
    # 101: (I4,4(3x,F8.5))
    return (
        f"{i:4d}"
        f"   {x:8.5f}"
        f"   {y:8.5f}"
        f"   {z:8.5f}"
        f"   {boy:8.5f}\n"
    )


def format_99(i: int, cano: int, boys: List[float]) -> str:
    # 99: (I4,2x,I4,10(3x,F8.5))
    s = f"{i:4d}  {cano:4d}"
    for v in boys:
        s += f"   {v:8.5f}"
    return s + "\n"


def read_slowmodes(path: str | Path, resno: int) -> List[List[float]]:
    # returns gnmboy[j][i] with j=0..9, i=0..resno-1
    gnmboy = [[0.0 for _ in range(resno)] for _ in range(10)]
    with open(path, "r", encoding="utf-8", errors="replace") as f:
        for i in range(resno):
            line = f.readline()
            if not line:
                raise RuntimeError(f"slowmodes ended early at residue {i+1}/{resno}")
            vals = [float(x) for x in line.strip().split()]
            if len(vals) < 10:
                vals = vals + [0.0] * (10 - len(vals))
            for j in range(10):
                gnmboy[j][i] = vals[j]
    return gnmboy


# ---------------------------- hinge finder -----------------------------------

def findhinge(infile: str | Path, resno: int) -> List[int]:
    """
    Mirror Fortran subroutine findhinge EXACTLY:
    - read first 13 chars, check (11:13) == 'nan'
    - backspace (i.e., re-process first line)
    - read formatted triples i j value sequentially
    - STOP as soon as j == resno for the first time
    - hinges where sign changes between mat(j) and mat(j+1)
    """
    path = Path(infile)

    with open(path, "r", encoding="utf-8", errors="replace") as f:
        first = f.readline()
        if not first:
            raise RuntimeError(f"Empty file: {infile}")

        charlin = (first.rstrip("\n") + " " * 13)[:13]  # A13
        if charlin[10:13].lower() == "nan":
            print(f"!!! nan entry in file !!! => {infile}")
            return []

        # "backspace": process first line again
        pending = [first] + f.readlines()

    mat = [0.0 for _ in range(resno + 1)]  # 1-based j
    reached_end = False

    for ln in pending:
        if not ln.strip():
            continue

        # Fortran 21: (1x,I4,1x,I4,1x,F6.1)
        # robust parse: try fixed slices first, then split
        s = ln.rstrip("\n")
        i = _as_int(s[1:5].strip(), 0)
        j = _as_int(s[6:10].strip(), 0)
        v = None
        if len(s) >= 11:
            v = _as_float(s[11:].strip(), 0.0)
        else:
            v = 0.0

        if j == 0:
            parts = ln.strip().split()
            if len(parts) >= 3:
                i = _as_int(parts[0], 0)
                j = _as_int(parts[1], 0)
                v = _as_float(parts[2], 0.0)

        if 1 <= j <= resno:
            mat[j] = float(v)

        if j == resno:
            reached_end = True
            break

    # Fortran assumes it hit j==resno; if not, proceed anyway with what we got
    hinges: List[int] = []
    for j in range(1, resno):
        a = mat[j]
        b = mat[j + 1]
        if (a > 0.0 and b < 0.0) or (a < 0.0 and b > 0.0):
            hinges.append(j)
            if len(hinges) >= 100:
                break

    return hinges


# ---------------------------- main pipeline ----------------------------------

def main(workdir: str = "."):
    wd = Path(workdir)

    rescale = _read_first_float(wd / "rescale")
    resno = _read_first_int(wd / "coordinates")

    file1 = "crosscorrslow1"
    file2 = "crosscorrslow2"

    alpha_lines = (wd / "alpha.cor").read_text(encoding="utf-8", errors="replace").splitlines()
    if len(alpha_lines) < resno:
        raise RuntimeError(f"alpha.cor has {len(alpha_lines)} lines, expected {resno}")

    # x,y,z,boy: [mode][res-1] mode=0..9, res=0..resno-1
    x = [[0.0 for _ in range(resno)] for _ in range(10)]
    y = [[0.0 for _ in range(resno)] for _ in range(10)]
    z = [[0.0 for _ in range(resno)] for _ in range(10)]
    boy = [[0.0 for _ in range(resno)] for _ in range(10)]

    # from alpha.cor: restyp(i), cano(i)
    restyp = [""] * resno
    cano = [0] * resno

    # from newcoordinat.mds: label, attyp, restyp, chtyp, ind, resex, xdist, ydist, zdist
    label = [""] * resno
    attyp = [""] * resno
    chtyp = [""] * resno
    ind = [0] * resno
    resex = [""] * resno
    xdist = [0.0] * resno
    ydist = [0.0] * resno
    zdist = [0.0] * resno

    # Open 1coor..10coor
    coor_fhs = [open(wd / f"{k}coor", "r", encoding="utf-8", errors="replace") for k in range(1, 11)]

    f_mds_in = open(wd / "newcoordinat.mds", "r", encoding="utf-8", errors="replace")
    f_mds_out = open(wd / "newcoordinat2.mds", "w", encoding="utf-8")
    f_len = open(wd / "anm_length", "w", encoding="utf-8")
    f_map = open(wd / "mapping.out", "w", encoding="utf-8")

    try:
        for i in range(resno):
            rt, cn = parse_alpha_line(alpha_lines[i])
            restyp[i] = rt
            cano[i] = cn

            # read coor lines for 10 modes
            for m in range(10):
                ln = coor_fhs[m].readline()
                if not ln:
                    raise RuntimeError(f"{m+1}coor ended early at residue {i+1}/{resno}")
                xi, yi, zi, bi = parse_coor_line(ln)
                x[m][i] = xi
                y[m][i] = yi
                z[m][i] = zi
                boy[m][i] = bi

            # read newcoordinat.mds line (format 100)
            mds_line = f_mds_in.readline()
            if not mds_line:
                raise RuntimeError(f"newcoordinat.mds ended early at residue {i+1}/{resno}")
            row = parse_mds100(mds_line)

            label[i] = row.label
            attyp[i] = row.attyp
            restyp[i] = row.restyp  # overwrite from file (as Fortran does)
            chtyp[i] = row.chtyp
            ind[i] = row.ind
            resex[i] = row.resex
            xdist[i] = row.x
            ydist[i] = row.y
            zdist[i] = row.z

            # Fortran: write (35,100) ... xdist,ydist,zdist  (NO occ in list!)
            f_mds_out.write(format_mds100(row, include_occ=False))

            # anm_length: write (31,99) i,cano(i),(boy(j,i),j=1,10)
            f_len.write(format_99(i + 1, cano[i], [boy[m][i] for m in range(10)]))

            # then boy squared
            for m in range(10):
                boy[m][i] = boy[m][i] ** 2

    finally:
        for fh in coor_fhs:
            fh.close()
        f_mds_in.close()
        f_mds_out.close()
        f_len.close()

    # ---------------- mapping: read slowmodes (gnmboy) ----------------
    gnmboy = read_slowmodes(wd / "slowmodes", resno=resno)  # [10][resno]

    # sums on i=8..resno-8 (Fortran inclusive). Python indices: 7..resno-9
    anmsum = [0.0] * 10
    gnmsum = [0.0] * 10
    for j in range(10):
        for i in range(7, resno - 8):
            gnmsum[j] += gnmboy[j][i]
            anmsum[j] += boy[j][i]

    # normalize
    anmboy = [[0.0 for _ in range(resno)] for _ in range(10)]
    for j in range(10):
        gden = gnmsum[j] if gnmsum[j] != 0.0 else 1.0
        aden = anmsum[j] if anmsum[j] != 0.0 else 1.0
        for i in range(resno):
            gnmboy[j][i] = gnmboy[j][i] / gden
            anmboy[j][i] = boy[j][i] / aden

    # error matrix e(1..2,1..10)
    e = [[0.0 for _ in range(10)] for _ in range(2)]
    for k in range(2):        # k=0..1 corresponds to Fortran k=1..2
        for j in range(10):   # j=0..9 corresponds to Fortran j=1..10
            acc = 0.0
            for i in range(7, resno - 8):
                acc += abs(gnmboy[k][i] - anmboy[j][i])
            e[k][j] = acc

    # Fortran selection rules (faithful translation)
    iii = 0  # Fortran iii=1
    for j in range(1, 10):  # Fortran j=2..10
        denom = e[0][iii]
        ratio = (e[0][j] / denom) if denom != 0.0 else float("inf")
        if ratio < 0.95:
            iii = j

    jjj = 0 if iii != 0 else 1  # Fortran: if iii!=1 jjj=1 else jjj=2
    for j in range(1, 10):      # Fortran j=2..10
        if j == iii:
            continue
        denom = e[1][jjj]
        ratio = (e[1][j] / denom) if denom != 0.0 else float("inf")
        if ratio < 0.95:
            jjj = j

    # mapping.out
    f_map.write(f"selection for GNM 1 : {iii+1}\n")
    f_map.write(f"selection for GNM 2 : {jjj+1}\n")
    f_map.write(" GNM 1st mode normalized parameters \n")
    f_map.write(
        "a1={:9.5f} a2={:9.5f} a3={:9.5f} a4={:9.5f} a5={:9.5f} a6={:9.5f}\n".format(
            e[0][0], e[0][1], e[0][2], e[0][3], e[0][4], e[0][5]
        )
    )
    f_map.write(" GNM 2nd mode normalized parameters \n")
    f_map.write(
        "a1={:9.5f} a2={:9.5f} a3={:9.5f} a4={:9.5f} a5={:9.5f} a6={:9.5f}\n".format(
            e[1][0], e[1][1], e[1][2], e[1][3], e[1][4], e[1][5]
        )
    )
    f_map.close()

    # Write coor*.mds12 and gnm*anmvector
    f41 = open(wd / "coor1.mds12", "w", encoding="utf-8")
    f42 = open(wd / "coor3.mds12", "w", encoding="utf-8")
    f43 = open(wd / "coor2.mds12", "w", encoding="utf-8")
    f44 = open(wd / "coor4.mds12", "w", encoding="utf-8")
    f45 = open(wd / "gnm1anmvector", "w", encoding="utf-8")
    f46 = open(wd / "gnm2anmvector", "w", encoding="utf-8")

    try:
        for i in range(resno):
            m1 = iii
            m2 = jjj

            base = AtomRow(
                label=label[i],
                attyp=attyp[i],
                restyp=restyp[i],
                chtyp=chtyp[i],
                ind=ind[i],
                resex=resex[i],
                x=xdist[i],
                y=ydist[i],
                z=zdist[i],
                occ=1.00,
            )

            # + m1
            f41.write(format_mds100(AtomRow(base.label, base.attyp, base.restyp, base.chtyp, base.ind, base.resex,
                                           base.x + x[m1][i] * rescale, base.y + y[m1][i] * rescale, base.z + z[m1][i] * rescale, 1.00),
                                    include_occ=True))
            # + m2
            f42.write(format_mds100(AtomRow(base.label, base.attyp, base.restyp, base.chtyp, base.ind, base.resex,
                                           base.x + x[m2][i] * rescale, base.y + y[m2][i] * rescale, base.z + z[m2][i] * rescale, 1.00),
                                    include_occ=True))
            # - m1
            f43.write(format_mds100(AtomRow(base.label, base.attyp, base.restyp, base.chtyp, base.ind, base.resex,
                                           base.x - x[m1][i] * rescale, base.y - y[m1][i] * rescale, base.z - z[m1][i] * rescale, 1.00),
                                    include_occ=True))
            # - m2
            f44.write(format_mds100(AtomRow(base.label, base.attyp, base.restyp, base.chtyp, base.ind, base.resex,
                                           base.x - x[m2][i] * rescale, base.y - y[m2][i] * rescale, base.z - z[m2][i] * rescale, 1.00),
                                    include_occ=True))

            # vectors (boy is already squared at this point, matching Fortran)
            f45.write(format_101(i + 1, x[m1][i], y[m1][i], z[m1][i], boy[m1][i]))
            f46.write(format_101(i + 1, x[m2][i], y[m2][i], z[m2][i], boy[m2][i]))

    finally:
        f41.close()
        f42.close()
        f43.close()
        f44.close()
        f45.close()
        f46.close()

    # ---------------- hinges from crosscorr files ----------------
    hinge1 = findhinge(wd / file1, resno=resno)
    hinge2 = findhinge(wd / file2, resno=resno)

    # Fortran 15: (3x,I4,1x,I4,A1,2x,A1)
    with open(wd / "hinges", "w", encoding="utf-8") as f:
        f.write(" ----> crosscorrelation : 1st slowest mode\n")
        for h in hinge1:
            idx = h - 1
            f.write(f"   {h:4d} {cano[idx]:4d}{resex[idx]:1s}  {chtyp[idx]:1s}\n")

        f.write(" ----> crosscorrelation : 2nd slowest mode\n")
        for h in hinge2:
            idx = h - 1
            f.write(f"   {h:4d} {cano[idx]:4d}{resex[idx]:1s}  {chtyp[idx]:1s}\n")


if __name__ == "__main__":
    main(".")

File: hingeprot/extract.f
       program hingebend
       integer hinge1(100),hinge2(100),n1,n2

       real anmsum(10),gnmsum(10)
       real x(10,5000),y(10,5000),z(10,5000)
       real e(10,10),f(10,10)
       real gnmboy(10,5000),boy(10,5000),anmboy(10,5000)
       real anmfark(10,5000),gnmfark(10,5000),rescale
       character*4 label(5000),attyp(5000)
       character*3 restyp(5000)
       character*2 chtyp(5000)
       integer ind(5000)
       character*1 resex(5000)
       real xdist(5000),ydist(5000),zdist(5000)
       integer hinge12(100),n12
       integer cano(5000),resno
       character*15 file1,file2,file5
       
       do i=1,100
       hinge1(i)=0
       hinge2(i)=0
       hinge12(i)=0
       end do
        
       n1=0
       n2=0
       n12=0

C13    format (4x,I4,3x,A3)
C14    format (2x,I4,4(3x,F8.5))
15    format (3x,I4,1x,I4,A1,2x,A1)

      open (unit=51,file='rescale')
      read(51,*)rescale
      file1='crosscorrslow1'
      file2='crosscorrslow2'
      file5='slowmodes'
      open (unit=19,file='coordinates')
      read (19,*) resno
      close(19)
      open (unit=20,file='alpha.cor')
      open (unit=21,file='1coor')
      open (unit=22,file='2coor')
      open (unit=23,file='3coor')
      open (unit=24,file='4coor')
      open (unit=25,file='5coor')
      open (unit=26,file='6coor')
      open (unit=27,file='7coor')
      open (unit=28,file='8coor')
      open (unit=29,file='9coor')
      open (unit=30,file='10coor')
      open (unit=31,file='anm_length')
      open (unit=32,file='newcoordinat.mds')
      open (unit=35,file='newcoordinat2.mds')
      open (unit=40,file='mapping.out')
      

97    format (17x,A3,2x,I4)
98    format (4x,4(3x,F8.5))
99    format (I4,2x,I4,10(3x,F8.5))
100   format (a4,8x,a4,1x,a3,1x,a1,i4,a1,f11.3,f8.3,f8.3,1x,f5.2)
105   format(a4,3x,I4,1x,a4,1x,a3,1x,a1,i4,a1,f11.3,f8.3,f8.3,
     :1x,f5.2,1x,i4,1x,f4.3,1x,f4.3,1x,f4.3)
101   format (I4,4(3x,F8.5))
      do j=1,10
        do i=1,resno
          boy(j,i)=0
        end do
      end do
      do i=1,resno
      read (20,97) restyp(i),cano(i)
      read (21,98) x(1,i),y(1,i),z(1,i),boy(1,i)
      read (22,98) x(2,i),y(2,i),z(2,i),boy(2,i)
      read (23,98) x(3,i),y(3,i),z(3,i),boy(3,i)
      read (24,98) x(4,i),y(4,i),z(4,i),boy(4,i)
      read (25,98) x(5,i),y(5,i),z(5,i),boy(5,i)
      read (26,98) x(6,i),y(6,i),z(6,i),boy(6,i)
      read (27,98) x(7,i),y(7,i),z(7,i),boy(7,i)
      read (28,98) x(8,i),y(8,i),z(8,i),boy(8,i)
      read (29,98) x(9,i),y(9,i),z(9,i),boy(9,i)
      read (30,98) x(10,i),y(10,i),z(10,i),boy(10,i)
      read (32,100) label(i),attyp(i),restyp(i),chtyp(i)
     :,ind(i),resex(i),xdist(i),ydist(i),zdist(i)
      write (35,100) label(i),attyp(i),restyp(i),chtyp(i)
     :,ind(i),resex(i),xdist(i),ydist(i),zdist(i)
      write (*,*)i
      write (31,99) i,cano(i),(boy(j,i),j=1,10)
       do j=1,10
       boy(j,i)=boy(j,i)**2
       enddo
      end do
      do i=20,32
      close (i)
      enddo

C*********   mapping ***************************************

16    format(4x,10(4x,F8.5))
      file5='slowmodes'
      open (unit=25,file=file5)
      do i=1,resno
      read (25,16) (gnmboy(j,i),j=1,10)
      end do
      close(25)

       do j=1,10
       anmsum(j)=0
       gnmsum(j)=0
       enddo

C     normalizaton sum
      do j=1,10
         do i=8,resno-8
          gnmsum(j)=gnmsum(j)+gnmboy(j,i)
          anmsum(j)=anmsum(j)+boy(j,i)
        enddo
      enddo

C      normazlize lengths
      do j=1,10
         do i=1,resno
          gnmboy(j,i)=gnmboy(j,i)/gnmsum(j)
          anmboy(j,i)=boy(j,i)/anmsum(j)
        enddo
      enddo


C     error calculation
      do i=1,2
       do j=1,10
        e(i,j)=0
        enddo
      enddo

      do k=1,2
       do j=1,10
        do i=8,resno-8
        e(k,j)=e(k,j)+abs(gnmboy(k,i)-anmboy(j,i))
        enddo
       enddo
      enddo


C      write (40,*) '*************',rescale
C     first mode
      iii=1
      do j=2,10
        if ((e(1,j)/e(1,iii)).lt.0.95) iii=j
      end do
C     second mode
      if (iii.ne.1) jjj=1
      if (iii.eq.1) jjj=2
      do j=2,10
      if (((e(2,j)/e(2,jjj)).lt.0.95).and.(j.ne.iii)) jjj=j
      end do
C     print results
      write (40,*) 'selection for GNM 1 : ',iii
      write (40,*) 'selection for GNM 2 : ',jjj
17    format ('a1=',F9.5,' a2=',F9.5,' a3=',F9.5,' a4=',F9.5,
     &' a5=',F9.5,' a6=',F9.5)

      write (40,*) ' GNM 1st mode normalized parameters '
      write (40,17) e(1,1),e(1,2),e(1,3),e(1,4),e(1,5),e(1,6)
      write (40,*) ' GNM 2nd mode normalized parameters '
      write (40,17) e(2,1),e(2,2),e(2,3),e(2,4),e(2,5),e(2,6)

      open (unit=41,file='coor1.mds12')
      open (unit=43,file='coor2.mds12')
      open (unit=42,file='coor3.mds12')
      open (unit=44,file='coor4.mds12')
      open (unit=45,file='gnm1anmvector')
      open (unit=46,file='gnm2anmvector')
      do i=1,resno
      write(41,100) label(i),attyp(i),restyp(i),chtyp(i)
     :,ind(i),resex(i),xdist(i)+x(iii,i)*rescale,
     :ydist(i)+y(iii,i)*rescale,zdist(i)+z(iii,i)*rescale,1.00
      write(42,100) label(i),attyp(i),restyp(i),chtyp(i)
     :,ind(i),resex(i),xdist(i)+x(jjj,i)*rescale,
     :ydist(i)+y(jjj,i)*rescale,zdist(i)+z(jjj,i)*rescale,1.00
      write(43,100) label(i),attyp(i),restyp(i),chtyp(i)
     :,ind(i),resex(i),xdist(i)-x(iii,i)*rescale,
     :ydist(i)-y(iii,i)*rescale,zdist(i)-z(iii,i)*rescale,1.00
      write(44,100) label(i),attyp(i),restyp(i),chtyp(i)
     :,ind(i),resex(i),xdist(i)-x(jjj,i)*rescale,
     :ydist(i)-y(jjj,i)*rescale,zdist(i)-z(jjj,i)*rescale,1.00
      write(45,101) i,x(iii,i),y(iii,i),z(iii,i),boy(iii,i)
      write(46,101) i,x(jjj,i),y(jjj,i),z(jjj,i),boy(jjj,i)
      enddo

C      write (6,*) ' GNM 1st mode differences '
C      write (6,17) f(1,1),f(1,2),f(1,3),f(1,4),f(1,5),f(1,6)
C      write (6,*) ' GNM 2nd mode differences '
C      write (6,17) f(2,1),f(2,2),f(2,3),f(2,4),f(2,5),f(2,6)

C***********************************************************

      call findhinge(file1,resno,hinge1,n1)
      call findhinge(file2,resno,hinge2,n2)
C      call findhinge(file3,resno,hinge12,n12)
      open (unit=18,file='hinges')

      write (18,*) '----> crosscorrelation : 1st slowest mode'
      do i=1,n1
      write (18,15) hinge1(i),cano(hinge1(i)),resex(hinge1(i)),
     &chtyp(hinge1(i))
      enddo


      write (18,*) '----> crosscorrelation : 2nd slowest mode'
      do i=1,n2
      write (18,15) hinge2(i),cano(hinge2(i)),resex(hinge2(i)),
     &chtyp(hinge2(i))
      enddo

      close(18)
      end
      
      
C*******************************************************************

      subroutine findhinge(infile,resno,hinge,n)

      character*15 infile
      character*13 charlin
      integer i,j,resno,n,hinge(100)
      real mat(5000)
      open (unit=11,file=infile)

       do j=1,resno
         mat(j)=0
       end do
       
21    format (1x,I4,1x,I4,1x,F6.1)
22    format (A13)
C23    format (10(1x,I3))


      read (11,22) charlin
      if (charlin(11:13).eq.'nan') goto 9999
      backspace (11)

2000  read (11,21)  i,j,mat(j)
      if (j.eq.resno) goto 2001
      goto 2000
2001  close(11)

      n=0
       do j=1,resno-1
        if ((mat(j).gt.0.and.mat(j+1).lt.0).or.
     :(mat(j).lt.0.and.mat(j+1).gt.0)) then
            n=n+1
            hinge(n)=j
          end if
        end do
C       print*,'. . . processing file : ',infile
C       write (6,23) (hinge(j),j=1,n)
C       print*,' '
       goto 10000
9999   print*,'!!! nan etnry in file !!! => ',infile
10000  continue
      end
         


